Add support for PHP 8, fix tests

ocean opened the pull request:
Targeting issue #55 

Still some test errors because of needing to change testing
for private/protected properties with PHPUnit 9 - I'm not sure how to test these properties now.

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean committed:
Add support for PHP 8, fix tests

Still some test errors because of needing to change testing
for private/protected properties with PHPUnit 9

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean committed:
Do not allow TravisCI failures for PHP 8 / nightly

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean committed:
Fix more broken tests, with help from #56

Thanks to the PR from @gennadigennadigennadi
I've incorporated those fixes into this PR to get PHPUnit 9 support.

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

froschdesign commented on test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php:
```diff
@@ -14,17 +14,19 @@
 use Laminas\ServiceManager\AbstractFactory\ReflectionBasedAbstractFactory;
 use Laminas\ServiceManager\Exception\ServiceNotFoundException;
 use PHPUnit\Framework\TestCase;
-use Prophecy\Prophecy\ObjectProphecy;
```
The class is still in use. See the DocBlock in line 24.

ocean replied:
Ok, added this back in.

ocean committed:
Improved some more tests but need some help

Can't figure out how to deal with the `getClass()` deprecations cleanly
in all instances in the files in this PR.

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean committed:
Reinstated missing class

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean commented:
The problematic area I need help with is around `/src/AbstractFactory/ReflectionBasedAbstractFactory.php:222`

snapshotpl commented:
@ocean this patch helps for me:

```
--- a/src/AbstractFactory/ReflectionBasedAbstractFactory.php
+++ b/src/AbstractFactory/ReflectionBasedAbstractFactory.php
@@ -217,13 +217,13 @@ class ReflectionBasedAbstractFactory implements AbstractFactoryInterface
             return [];
         }
 
-        if ($parameter->hasType() && ! $parameter->getType()->isBuiltin()) {
+        if (! $parameter->hasType() || ($parameter->hasType() && $parameter->getType()->isBuiltin())) {
             if (! $parameter->isDefaultValueAvailable()) {
                 throw new ServiceNotFoundException(sprintf(
                     'Unable to create service "%s"; unable to resolve parameter "%s" '
                     . 'to a class, interface, or array type',
                     $requestedName,
-                    $parameter->getType()
+                    $parameter->getName()
                 ));
             }
```

ocean committed:
Fix failing parameter type test  with help from @snapshotpl's patch

Thanks https://github.com/snapshotpl

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean committed:
Fix failing test: $options will be ignored

Perhaps test should not look at this, given it's ignored by the class?

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean commented:
@froschdesign This is ready for review now 😃

froschdesign commented on src/AbstractFactory/ReflectionBasedAbstractFactory.php:
```diff
@@ -249,4 +250,30 @@ private function resolveParameter(ReflectionParameter $parameter, ContainerInter
+     * @return string|null
+     *   The parameter's class name or NULL if the parameter is not a class.
+     */
+    public static function getParameterClassName(ReflectionParameter $parameter)
```
This method is declared as static, but is never used in this form. And I see no reason to extend the public API here.
Please set the visibility to private and remove `static`.

froschdesign commented on src/Tool/FactoryCreator.php:
```diff
@@ -80,6 +80,32 @@ public function createFactory($className)
+     * @return string|null
+     *   The parameter's class name or NULL if the parameter is not a class.
+     */
+    public static function getParameterClassName(ReflectionParameter $parameter)
```
The same here.

froschdesign commented on test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php:
```diff
@@ -171,7 +174,7 @@ public function testFactoryCanSupplyAMixOfParameterTypes()
 
         self::assertEquals($config, $instance->config);
-        self::assertEquals([], $instance->options);
+        self::assertEquals(null, $instance->options);
```
Does this change mean that there was an error before?

ocean replied:
Yes, there was an error in the tests when I ran them locally. I thought I saw a comment something about `$options` always being treated as `null`, but now I can't find it. I'll revert this change and then you can see the error that shows up in Travis.

ocean replied:
@froschdesign Yep, in the test the `options` instance property is set to `null`, not to an empty array, see:
https://github.com/laminas/laminas-servicemanager/blob/3.5.x/test/AbstractFactory/TestAsset/ClassWithMixedConstructorParameters.php#L24

So either the assertion or the test class constructor needs changing I guess.

froschdesign reviewed (changes requested):
Thanks for your contribution! 👍 

Two changes are necessary and I have added a question.

ocean committed:
Change these public static methods to private

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean committed:
Revert test $instance->options null check change

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

ocean committed:
Ignore PHPUnit result cache file

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

boesing commented:
@ocean Is there anything we can help with? Feel free to ping me.

**Update: I've merged #56 which already had phpunit 9 fixes and thus your tests might pass after rebasing against 3.5.x branch 👍**

boesing commented:
Oh, I wasnt aware that there was already #51 which added PHP 8.0 compatibility back in july.
I'll close this in favor of #51 as its tests are already passing.

FabianKoestring commented:
Could you tag a new release for this feature?

Ocramius commented:
@FabianKoestring done

Ocramius commented:
@FabianKoestring beware: due to #59, this has been reverted for now, and will therefore be delayed.
